### PR TITLE
Bug fix when `nstates=1`; when `expects is None`; small changes to plotting

### PR DESCRIPTION
**Pull request by cyrus-baker**
1. fix bug when plot_expects when expects is None; 
2. fix dim calculation in PropagatorInfidelity; 
3. fix bug when initial state is only one in CoherentInfidelity

### CODE REVIEW

File: qontrol/cost.py [hunk 1]
```diff
@@ -375,16 +375,16 @@ def __call__(
     ) -> tuple[tuple[Array, Array]]:
         overlaps = self.target_states.dag() @ result.final_state
         if not isket(result.final_state):
-            overlaps = overlaps.trace()
+            overlaps = overlaps.trace(axis1=-1, axis2=-2)
```
Comment (dkweiss31) [inline qontrol/cost.py]: I'm not sure this is necessary, see [here](https://github.com/dynamiqs/dynamiqs/blob/main/dynamiqs/qarrays/dense_qarray.py#L87-L88) for the function this calls which already specifies the last two axes

File: qontrol/cost.py [hunk 2]
```diff
@@ -375,16 +375,16 @@ def __call__(
...
             # average over states before squaring: for density matrices this doesn't do
             # anything different from the incoherent definition of the infidelity, since
             # the trace is always real and positive. Included here only for
             # completeness.
-            overlaps_avg = jnp.mean(jnp.squeeze(overlaps), axis=-1)
+            overlaps_avg = jnp.mean(jnp.squeeze(overlaps, axis=[-1, -2]), axis=-1)
```
Comment (dkweiss31) [inline qontrol/cost.py]: Good catch!

File: qontrol/cost.py [hunk 3]
```diff
@@ -401,7 +401,7 @@ def __call__(
         H: TimeQArray,  # noqa ARG002
         parameters: dict | Array,  # noqa ARG002
     ) -> tuple[tuple[Array, Array]]:
-        dim = jnp.prod(*result.final_propagator.dims)
+        dim = jnp.prod(jnp.array(result.final_propagator.dims))
```
Comment (dkweiss31) [inline qontrol/cost.py:404]: Would the simpler
```suggestion
        dim = jnp.prod(result.final_propagator.dims)
```
work?

File: qontrol/plot.py [hunk 4]
```diff
@@ -100,9 +100,10 @@ def plot_expects(
     ax.set_facecolor('none')
     tsave = model.tsave_function(parameters)
     # plot all expectation values by default
-    expect_idxs = np.ndindex(*expects.shape[:-1])
-    for expect_idx in expect_idxs:
-        ax.plot(tsave, np.real(expects[tuple(expect_idx)]))
+    if expects is not None:
+        expect_idxs = np.ndindex(*expects.shape[:-1])
+        for expect_idx in expect_idxs:
+            ax.plot(tsave, np.real(expects[tuple(expect_idx)]))
```
Comment (dkweiss31) [inline qontrol/plot.py:109]: This isn't going to be the ideal solution, since it will leave a blank panel (regardless, good catch!). Maybe instead we could use different default plotters for mesolve, sesolve versus mepropogator, sepropogator?

**Review by dkweiss31 — CHANGES_REQUESTED** [review]
Thanks for this PR @littlebaker! I've left some comments. I'm wondering if you might be able to adjust your solution when `expects is None` to avoid just having an empty plot. Could we use different default plotters for the propogator functionality? Additionally, you've uncovered a few bugs, and I'm wondering if you can add test coverage for those use cases so that they shouldn't show up again?

File: qontrol/cost.py (hunk 1 above)
Comment (cyrus-baker) [reply qontrol/cost.py]: Oh yes, this bug is caused by `squeeze`, not by `trace`

File: qontrol/cost.py (hunk 3 above)
Comment (cyrus-baker) [reply qontrol/cost.py:404]: No, `result.final_propagator.dims` is a tuple. However, jax require only `jnp.ndarray` and do not accept implict convertion

File: qontrol/plot.py (hunk 4 above)
Comment (cyrus-baker) [reply qontrol/plot.py:109]: You are right, this is just a temporary solution. I will see if there is a better solution. Maybe instead of using propagators for "expectations", it is better to use "states" and calculate their evolution. But this may lead to changes in semantics and affect the API.

**Comment by cyrus-baker** [comment]
Sure, I will remove the empty plot when `expects is None`. As for the plotters functionality, I think we may need further discussion.

**Comment by cyrus-baker** [comment]
Sure, I will add some test code

File: qontrol/optimize.py [hunk 5]
```diff
@@ -104,6 +104,18 @@ def optimize(
...
+        if (
+            model.exp_ops is not None
+            and len(model.exp_ops) > 0
+            and not isinstance(model, SEPropagatorModel)
+            and not isinstance(model, MEPropagatorModel)
+        ):
+            plotter = DefaultPlotter()
+        else:
+            plotter = Plotter([plot_fft, plot_controls])
+
```
Comment (dkweiss31) [inline qontrol/optimize.py]: I like what you're going for here. I think what this is pointing to is that we need one more layer of abstraction. The base `Model` class shouldn't have `exp_ops` as an attribute, since the propagator models don't have it. I'm thinking of something like
```python
class Model(eqx.Module):
    H_function: callable
    tsave_function: callable

    def __call__(
            self,
            parameters: Array | dict,
            method: Method = Tsit5(),  # noqa B008
            gradient: Gradient | None = None,
            options: dq.Options = dq.Options(),  # noqa B008
    ) -> tuple[Result, TimeQArray]:
        raise NotImplementedError


class SolveModel(Model):
    exp_ops: Array | None


class PropagatorModel(Model):
    pass
```

Would you be able to make these changes to the model file? Then the check above could become much simpler.

**Comment by dkweiss31** [comment]
I think we're getting close! One quick thing: there are a lot of .pyc files that are part of this PR: can you remove them?

File: qontrol/optimize.py (hunk 5 above)
Comment (cyrus-baker) [reply qontrol/optimize.py]: Sure, that will be a good idea.

**Comment by cyrus-baker** [comment]
Ok, sure. I will remove them.

**Comment by cyrus-baker** [comment]
Sorry, wrong operation, I did not finished.

**Comment by cyrus-baker** [comment]
```python
def plot_controls(
    ax: Axes, _expects: Array | None, model: Model, parameters: Array | dict
) -> Axes:
    """Plot the Hamiltonian prefactors, usually corresponding to controls."""
    ax.set_facecolor('none')
    H = model.H_function(parameters)
    tsave = model.tsave_function(parameters)
    controls = get_controls(H, tsave)
    H_labels = [f'$H_{idx}$' for idx in range(len(controls))]
    for idx, control in enumerate(controls):
        ax.plot(tsave, np.real(control) / 2 / np.pi, label=H_labels[idx])
    ax.legend(loc='lower right', framealpha=0.0)
    ax.set_ylabel('pulse amplitude [GHz]')
    ax.set_xlabel('time [ns]')
    return ax
```
why `np.real(control) / 2 / np.pi` has `/ 2 / np.pi` ? I think 2 pi should be multiplied in Hamiltonians?

**Comment by dkweiss31** [comment]
Right, this is a mostly arbitrary choice that the controls included the 2pis, so that I needed to divide by 2pi when plotting. You are right though that not everyone will do this and it can be unexpected for users, so I'm happy to change this default behavior (though one can of course use their own custom `Plotter`

**Comment by dkweiss31** [comment]
@littlebaker is this PR ready for my review? (I do still see a .json and .pyc file that shouldn't be included)

**Comment by cyrus-baker** [comment]
Oh, sorry for that. I will check it again.

**Review by dkweiss31 — APPROVED** [review]
Thanks @littlebaker ! Everything LGTM. I will let the ci run and then we can merge :). Congrats on your first contribution!!
